File: src/main/java/cz/vhromada/catalog/rest/controller/GenreController.java
package cz.vhromada.catalog.rest.controller;

import java.util.List;

import cz.vhromada.catalog.entity.Genre;
import cz.vhromada.catalog.facade.GenreFacade;
import cz.vhromada.validation.result.Result;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.util.Assert;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * A class represents controller for genres.
 *
 * @author Vladimir Hromada
 */
@RestController("genreController")
@RequestMapping("/catalog/genres")
public class GenreController extends AbstractCatalogController {

    /**
     * Facade for genres
     */
    private GenreFacade genreFacade;

    /**
     * Creates a new instance of GenreController.
     *
     * @param genreFacade facade for genres
     * @throws IllegalArgumentException if facade for genres is null
     */
    @Autowired
    public GenreController(final GenreFacade genreFacade) {
        Assert.notNull(genreFacade, "Facade for genres mustn't be null.");

        this.genreFacade = genreFacade;
    }

    /**
     * Creates new data.
     *
     * @return result
     */
    @PostMapping("/new")
    public ResponseEntity<Result<Void>> newData() {
        return processResult(genreFacade.newData());
    }

    /**
     * Returns list of genres.
     *
     * @return result with list of genres
     */
    @GetMapping
    public ResponseEntity<Result<List<Genre>>> getGenres() {
        return processResult(genreFacade.getAll());
    }

    /**
     * Returns genre with ID or null if there isn't such genre.
     * <br>
     * Validation errors:
     * <ul>
     * <li>ID is null</li>
     * </ul>
     *
     * @param id ID
     * @return result with genre or validation errors
     */
    @GetMapping("/{id}")
    public ResponseEntity<Result<Genre>> getGenre(@PathVariable("id") final Integer id) {
        return processResult(genreFacade.get(id));
    }

    /**
     * Adds genre. Sets new ID and position.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Genre is null</li>
     * <li>ID isn't null</li>
     * <li>Position isn't null</li>
     * <li>Name is null</li>
     * <li>Name is empty string</li>
     * </ul>
     *
     * @param genre genre
     * @return result with validation errors
     */
    @PutMapping("/add")
    public ResponseEntity<Result<Void>> add(@RequestBody final Genre genre) {
        return processResult(genreFacade.add(genre), HttpStatus.CREATED);
    }

    /**
     * Updates genre.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Genre is null</li>
     * <li>ID is null</li>
     * <li>Position is null</li>
     * <li>Name is null</li>
     * <li>Name is empty string</li>
     * <li>Genre doesn't exist in data storage</li>
     * </ul>
     *
     * @param genre new value of genre
     * @return result with validation errors
     */
    @PostMapping("/update")
    public ResponseEntity<Result<Void>> update(@RequestBody final Genre genre) {
        return processResult(genreFacade.update(genre));
    }

    /**
     * Removes genre.
     * <br>
     * Validation errors:
     * <ul>
     * <li>ID is null</li>
     * <li>Genre doesn't exist in data storage</li>
     * </ul>
     *
     * @param id ID
     * @return result with validation errors
     */
    @DeleteMapping("/remove/{id}")
    public ResponseEntity<Result<Void>> remove(@PathVariable("id") final Integer id) {
        final Genre genre = new Genre();
        genre.setId(id);
        return processResult(genreFacade.remove(genre));
    }

    /**
     * Duplicates genre.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Genre is null</li>
     * <li>ID is null</li>
     * <li>Genre doesn't exist in data storage</li>
     * </ul>
     *
     * @param genre genre
     * @return result with validation errors
     */
    @PostMapping("/duplicate")
    public ResponseEntity<Result<Void>> duplicate(@RequestBody final Genre genre) {
        return processResult(genreFacade.duplicate(genre), HttpStatus.CREATED);
    }

    /**
     * Moves genre in list one position up.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Genre is null</li>
     * <li>ID is null</li>
     * <li>Genre can't be moved up</li>
     * <li>Genre doesn't exist in data storage</li>
     * </ul>
     *
     * @param genre genre
     * @return result with validation errors
     */
    @PostMapping("/moveUp")
    public ResponseEntity<Result<Void>> moveUp(@RequestBody final Genre genre) {
        return processResult(genreFacade.moveUp(genre));
    }

    /**
     * Moves genre in list one position down.
     * <br>
     * Validation errors:
     * <ul>
     * <li>Genre is null</li>
     * <li>ID is null</li>
     * <li>Genre can't be moved down</li>
     * <li>Genre doesn't exist in data storage</li>
     * </ul>
     *
     * @param genre genre
     * @return result with validation errors
     */
    @PostMapping("/moveDown")
    public ResponseEntity<Result<Void>> moveDown(@RequestBody final Genre genre) {
        return processResult(genreFacade.moveDown(genre));
    }

    /**
     * Updates positions.
     *
     * @return result
     */
    @PostMapping("/updatePositions")
    public ResponseEntity<Result<Void>> updatePositions() {
        return processResult(genreFacade.updatePositions());
    }

}

File: src/main/kotlin/cz/vhromada/catalog/rest/controller/GenreController.kt
package cz.vhromada.catalog.rest.controller

import cz.vhromada.catalog.entity.Genre
import cz.vhromada.catalog.facade.GenreFacade
import cz.vhromada.common.web.controller.AbstractController
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController

/**
 * A class represents controller for genres.
 *
 * @author Vladimir Hromada
 */
@RestController("genreController")
@RequestMapping("/catalog/genres")
class GenreController(private val genreFacade: GenreFacade) : AbstractController() {

    /**
     * Returns list of genres.
     *
     * @return list of genres
     */
    @GetMapping
    fun getGenres(): List<Genre> {
        return processResult(genreFacade.getAll())!!
    }

    /**
     * Creates new data.
     */
    @PostMapping("/new")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun newData() {
        processResult(genreFacade.newData())
    }

    /**
     * Returns genre with ID or null if there isn't such genre.
     *
     * @param id ID
     * @return genre with ID or null if there isn't such genre
     */
    @GetMapping("/{id}")
    fun getGenre(@PathVariable("id") id: Int): Genre? {
        return processResult(genreFacade.get(id))
    }

    /**
     * Adds genre. Sets new ID and position.
     * <br></br>
     * Validation errors:
     *
     *  * ID isn't null
     *  * Position isn't null
     *  * Name is null
     *  * Name is empty string
     *
     * @param genre genre
     */
    @PutMapping("/add")
    @ResponseStatus(HttpStatus.CREATED)
    fun add(@RequestBody genre: Genre) {
        processResult(genreFacade.add(genre))
    }

    /**
     * Updates genre.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Position is null
     *  * Name is null
     *  * Name is empty string
     *  * Genre doesn't exist in data storage
     *
     * @param genre new value of genre
     */
    @PostMapping("/update")
    fun update(@RequestBody genre: Genre) {
        processResult(genreFacade.update(genre))
    }

    /**
     * Removes genre.
     * <br></br>
     * Validation errors:
     *
     *  * Genre doesn't exist in data storage
     *
     * @param id ID
     */
    @DeleteMapping("/remove/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun remove(@PathVariable("id") id: Int) {
        val genre = Genre(id = id, name = null, position = null)
        processResult(genreFacade.remove(genre))
    }

    /**
     * Duplicates genre.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Genre doesn't exist in data storage
     *
     * @param genre genre
     */
    @PostMapping("/duplicate")
    @ResponseStatus(HttpStatus.CREATED)
    fun duplicate(@RequestBody genre: Genre) {
        processResult(genreFacade.duplicate(genre))
    }

    /**
     * Moves genre in list one position up.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Genre can't be moved up
     *  * Genre doesn't exist in data storage
     *
     * @param genre genre
     */
    @PostMapping("/moveUp")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveUp(@RequestBody genre: Genre) {
        processResult(genreFacade.moveUp(genre))
    }

    /**
     * Moves genre in list one position down.
     * <br></br>
     * Validation errors:
     *
     *  * ID is null
     *  * Genre can't be moved down
     *  * Genre doesn't exist in data storage
     *
     * @param genre genre
     */
    @PostMapping("/moveDown")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun moveDown(@RequestBody genre: Genre) {
        processResult(genreFacade.moveDown(genre))
    }

    /**
     * Updates positions.
     */
    @PostMapping("/updatePositions")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun updatePositions() {
        processResult(genreFacade.updatePositions())
    }

}
